.github/scripts/binaryCompatibility.py

```diff
@@ -0,0 +1,76 @@
+#  Copyright Amazon.com Inc. or its affiliates.
+#  SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import json
+import os
+import xml.etree.ElementTree as ET
+
+from agithub.GitHub import GitHub
+
+
+def findall_recursive(node, element):
+    for item in node.findall(element):
+        yield item
+    for item in list(node):
+        yield from findall_recursive(item, element)
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--input', type=argparse.FileType('r'), help='Input file to parse')
+    parser.add_argument('--token', type=str, help='GitHub token')
+    args = parser.parse_args()
+
+    file = args.input
+
+    my_body_dedupe = "Produced by binaryCompatability.py"
+    body = f"Binary incompatibility detected for commit {os.getenv('GITHUB_SHA')}.\n" \
+           f" See the uploaded artifacts from the action for details. You must bump the version number.\n\n"
+
+    incompatible = False
+    tree = ET.parse(file)
+    for clazz in tree.getroot().findall("./classes/class"):
+        any_incompatible = False
+        body += f"{clazz.attrib['fullyQualifiedName']} is "
+        if clazz.attrib['binaryCompatible'] == "false":
+            body += "binary incompatible"
+            incompatible = True
+            any_incompatible = True
+        if clazz.attrib['sourceCompatible'] == "false":
+            if any_incompatible:
+                body += " and is "
+            body += "source incompatible"
+            any_incompatible = True
+        if any_incompatible:
+            body += f" because of " \
+                    f"{', '.join({x.text for x in findall_recursive(clazz, 'compatibilityChanges/compatibilityChange')})}"
+        else:
+            body += "fully compatible"
+        body += "\n"
+    body += "\n" + my_body_dedupe
+
+    token = args.token
+    pr = json.load(open(os.getenv("GITHUB_EVENT_PATH"), 'r'))["pull_request"]["number"]
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *open*. The resource is allocated by call *builtins.open*. Execution paths that do not contain closure statements were detected. To prevent this resource leak, close the object returned by *open()* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).